src/Sarif/Visitors/PartitioningVisitor.cs

```diff
@@ -0,0 +1,454 @@
+                && node.Index < currentRunArtifacts.Count)
+            {
+                // Does this artifactLocation occur in the context of a result...
+                if (inResult)
+                {
+                    // If so, does this result belong to any partition?
+                    if (currentPartitionValue != null)
+                    {
+                        // Remap this index if it isn't already remapped.
+                        Dictionary<T, HashSet<int>> partitionArtifactIndicesDictionary = partitionResultArtifactIndicesDictionaries[currentRunIndex];
+                        if (!partitionArtifactIndicesDictionary.ContainsKey(currentPartitionValue))
+                        {
+                            partitionArtifactIndicesDictionary.Add(currentPartitionValue, new HashSet<int>());
+                        }
+
+                        artifactIndices = partitionArtifactIndicesDictionary[currentPartitionValue];
+                    }
+                }
+                // ... or outside of any result?
+                else
+                {
+                    artifactIndices = globalArtifactIndexSets[currentRunIndex];
+                }
+
+                // Either way, bring the artifact and all its ancestors into the appropriate
+                // list of indices.
+                if (artifactIndices != null)
+                {
+                    // It's safe to add the same item to a hash set multiple times, but
+                    // we check whether we've already seen this item anyway to avoid walking
+                    // the ancestor chain if we don't need to.
+                    if (!artifactIndices.Contains(node.Index))
+                    {
+                        artifactIndices.Add(node.Index);
+
+                        Artifact artifact = currentRunArtifacts[node.Index];
+                        while (artifact.ParentIndex >= 0)
+                        {
+                            artifactIndices.Add(artifact.ParentIndex);
+                            artifact = currentRunArtifacts[artifact.ParentIndex];
+                        }
+                    }
+                }
+            }
+
+            return base.VisitArtifactLocation(node);
+        }
+
+        private void CreatePartitionLogDictionary()
+        {
+            partitionLogDictionary = partitionValues.ToDictionary(
+                keySelector: pv => pv,
+                elementSelector: CreatePartitionLog);
+        }
+
+        private SarifLog CreatePartitionLog(T partitionValue)
+        {
+            SarifLog partitionLog;
+            if (deepClone)
+            {
+                // Save time and space by not cloning the runs unless and until necessary. We will
+                // only need to clone the runs that have results in this partition.
+                IList<Run> originalRuns = originalLog.Runs;
+                originalLog.Runs = null;
+
+                partitionLog = originalLog.DeepClone();
```

Comment:
Is this actually different than new SarifLog(originalLog)? Are there any properties outside the Runs which need DeepCloning?

Reply:
Yes, `sarifLog.inlineExternalProperties`.

---
In reply to: [357431410](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357431410) [](ancestors = 357431410)

Reply:
(Even if there weren't any such properties, it's good future-proofing against the day when there might be some.)

---
In reply to: [357440439](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357440439) [](ancestors = 357440439,357431410)